src/nnue/layers/affine_transform.h

```diff
@@ -74,113 +74,578 @@ namespace Eval::NNUE::Layers {
+            product1 = _mm256_madd_epi16(product1, kOnes256);
+            sum1 = _mm256_add_epi32(sum1, product1);
+
+            __m256i product2 = _mm256_maddubs_epi16(in, _mm256_load_si256(&row2[0]));
+            product2 = _mm256_madd_epi16(product2, kOnes256);
+            sum2 = _mm256_add_epi32(sum2, product2);
+
+            __m256i product3 = _mm256_maddubs_epi16(in, _mm256_load_si256(&row3[0]));
+            product3 = _mm256_madd_epi16(product3, kOnes256);
+            sum3 = _mm256_add_epi32(sum3, product3);
+
+            const __m128i out = m256_haddx4(sum0, sum1, sum2, sum3, bias);
+            _mm_store_si128(outptr, out);
+          }
+          else
+          {
+            assert(false);
+          }
+        }
+      }
+      else if constexpr (kOutputDimensions == 1)
+      {
+        static_assert(kNumChunks == 0);
+
+        const auto row = reinterpret_cast<const __m256i*>(&weights_[0]);
+
+        __m256i product = _mm256_maddubs_epi16(_mm256_loadA_si256(&input_vector256[0]), _mm256_load_si256(&row[0]));
+        product = _mm256_madd_epi16(product, kOnes256);
+
+        output[0] = m256_hadd(product, biases_[0]);
+      }
+      else
+      {
+        exit(1);
+      }
+      return output;
+
+#elif defined (USE_AVX2) && defined (USE_VNNI)
+
+      const auto output = reinterpret_cast<OutputType*>(buffer);
 
-  #elif defined(USE_NEON)
       constexpr IndexType kNumChunks = kPaddedInputDimensions / kSimdWidth;
-      const auto input_vector = reinterpret_cast<const int8x8_t*>(input);
-  #endif
+      const auto input_vector = reinterpret_cast<const __m256i*>(input);
 
-      for (IndexType i = 0; i < kOutputDimensions; ++i) {
-        const IndexType offset = i * kPaddedInputDimensions;
+      if constexpr (kOutputDimensions % 4 == 0)
+      {
+        for (IndexType i = 0; i < kOutputDimensions; i += 4) {
+          const IndexType offset0 = (i + 0) * kPaddedInputDimensions;
+          const IndexType offset1 = (i + 1) * kPaddedInputDimensions;
+          const IndexType offset2 = (i + 2) * kPaddedInputDimensions;
+          const IndexType offset3 = (i + 3) * kPaddedInputDimensions;
+
+          const __m128i bias = _mm_load_si128(reinterpret_cast<const __m128i*>(&biases_[i]));
+          __m128i* outptr = reinterpret_cast<__m128i*>(&output[i]);
+
+          __m256i sum0 = _mm256_setzero_si256();
+          __m256i sum1 = _mm256_setzero_si256();
+          __m256i sum2 = _mm256_setzero_si256();
+          __m256i sum3 = _mm256_setzero_si256();
+
+          const auto row0 = reinterpret_cast<const __m256i*>(&weights_[offset0]);
+          const auto row1 = reinterpret_cast<const __m256i*>(&weights_[offset1]);
+          const auto row2 = reinterpret_cast<const __m256i*>(&weights_[offset2]);
+          const auto row3 = reinterpret_cast<const __m256i*>(&weights_[offset3]);
+
+          IndexType j = 0;
+          for (; j < kNumChunks - 1; j += 2) {
```

Comment:
It is safe to leave it to the compiler to unroll the loop.

Reply:
yea, I'm tryout out snippets on godbolt now to see what can be simplified.